File: frontend/package/src/Score/Key/KeySignatures.ts
import Accidental from "../Pitch/Accidental";
import ScaleValue from "../Pitch/ScaleValue";
import KeySignature from "./KeySignature";
import Mode from './Mode';

class KeySignatures {
  static readonly SEVEN_FLAT: KeySignature = new KeySignature(
    [Accidental.FLAT, Accidental.FLAT, Accidental.FLAT, Accidental.FLAT, Accidental.FLAT , Accidental.FLAT, Accidental.FLAT], 
    ScaleValue.C)

  static readonly SIX_FLAT: KeySignature = new KeySignature(
      [Accidental.FLAT, Accidental.FLAT, Accidental.FLAT, Accidental.FLAT, Accidental.FLAT , Accidental.FLAT, Accidental.NAT], 
      ScaleValue.G)

  static readonly FIVE_FLAT: KeySignature = new KeySignature(
    [Accidental.FLAT, Accidental.FLAT, Accidental.NAT, Accidental.FLAT, Accidental.FLAT , Accidental.FLAT, Accidental.NAT], 
    ScaleValue.D)

  static readonly FOUR_FLAT: KeySignature = new KeySignature(
    [Accidental.FLAT, Accidental.FLAT, Accidental.NAT, Accidental.FLAT, Accidental.FLAT , Accidental.NAT, Accidental.NAT], 
    ScaleValue.A)

  static readonly THREE_FLAT: KeySignature = new KeySignature(
    [Accidental.FLAT, Accidental.NAT, Accidental.NAT, Accidental.FLAT, Accidental.FLAT , Accidental.NAT, Accidental.NAT], 
    ScaleValue.E)

  static readonly TWO_FLAT: KeySignature = new KeySignature(
    [Accidental.FLAT, Accidental.NAT, Accidental.NAT, Accidental.FLAT, Accidental.NAT , Accidental.NAT, Accidental.NAT], 
    ScaleValue.B)

  static readonly ONE_FLAT: KeySignature = new KeySignature(
    [Accidental.NAT, Accidental.NAT, Accidental.NAT, Accidental.FLAT, Accidental.NAT , Accidental.NAT, Accidental.NAT], 
    ScaleValue.F)

  static readonly NO_ACCIDENTAL: KeySignature = new KeySignature(
    [Accidental.NAT, Accidental.NAT, Accidental.NAT, Accidental.NAT, Accidental.NAT , Accidental.NAT, Accidental.NAT], 
    ScaleValue.C)
  
  static readonly ONE_SHARP: KeySignature = new KeySignature(
    [Accidental.NAT, Accidental.NAT, Accidental.NAT, Accidental.NAT, Accidental.NAT , Accidental.NAT, Accidental.SHARP], 
    ScaleValue.G)
 
  static readonly TWO_SHARP: KeySignature = new KeySignature(
    [Accidental.NAT, Accidental.NAT, Accidental.SHARP, Accidental.NAT, Accidental.NAT , Accidental.NAT, Accidental.SHARP], 
    ScaleValue.D)

  static readonly THREE_SHARP: KeySignature = new KeySignature(
    [Accidental.NAT, Accidental.NAT, Accidental.SHARP, Accidental.NAT, Accidental.NAT , Accidental.SHARP, Accidental.SHARP], 
    ScaleValue.A)

  static readonly FOUR_SHARP: KeySignature = new KeySignature(
    [Accidental.NAT, Accidental.SHARP, Accidental.SHARP, Accidental.NAT, Accidental.NAT , Accidental.SHARP, Accidental.SHARP], 
    ScaleValue.E)

  static readonly FIVE_SHARP: KeySignature = new KeySignature(
    [Accidental.NAT, Accidental.SHARP, Accidental.SHARP, Accidental.NAT, Accidental.SHARP , Accidental.SHARP, Accidental.SHARP], 
    ScaleValue.B)

  static readonly SIX_SHARP: KeySignature = new KeySignature(
    [Accidental.SHARP, Accidental.SHARP, Accidental.SHARP, Accidental.NAT, Accidental.SHARP , Accidental.SHARP, Accidental.SHARP], 
    ScaleValue.F)

  static readonly SEVEN_SHARP: KeySignature = new KeySignature(
    [Accidental.SHARP, Accidental.SHARP, Accidental.SHARP, Accidental.SHARP, Accidental.SHARP , Accidental.SHARP, Accidental.SHARP], 
    ScaleValue.C)

  //~~~

  static readonly ALL_KEY_SIGNATURES: Array<KeySignature> = 
    [KeySignatures.SEVEN_FLAT, KeySignatures.SIX_FLAT, KeySignatures.FIVE_FLAT, KeySignatures.FOUR_FLAT, KeySignatures.THREE_FLAT, KeySignatures.TWO_FLAT, KeySignatures.ONE_FLAT, KeySignatures.NO_ACCIDENTAL, 
      KeySignatures.ONE_SHARP, KeySignatures.TWO_SHARP, KeySignatures.THREE_SHARP, KeySignatures.FOUR_SHARP, KeySignatures.FIVE_SHARP, KeySignatures.SIX_SHARP, KeySignatures.SEVEN_SHARP]

  static fromEncoding(sv : ScaleValue, a : Accidental, m : Mode): KeySignature {
    for (let k of this.ALL_KEY_SIGNATURES) {
      let noteNameScale = k.getScaleLetterNames(k.ionianTonic);

      if (noteNameScale[m] == sv && k.accidentals[m] == a) {
        return k;
      }
    }
    return this.NO_ACCIDENTAL;
  }

  //~~~

  private constructor() {
  }
  
}

export default KeySignatures;

File: server/src/main/java/edu/brown/cs/counterpointer/notenotation/key/KeySignatures.java
package edu.brown.cs.counterpointer.notenotation.key;

import edu.brown.cs.counterpointer.notenotation.exceptions.InvalidScaleException;
import edu.brown.cs.counterpointer.notenotation.pitch.Accidental;
import edu.brown.cs.counterpointer.notenotation.pitch.ScaleValue;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

import static edu.brown.cs.counterpointer.notenotation.pitch.Accidental.FLAT;
import static edu.brown.cs.counterpointer.notenotation.pitch.Accidental.NATURAL;
import static edu.brown.cs.counterpointer.notenotation.pitch.Accidental.SHARP;

/**
 * A utility class storing all key signatures used and accepted during the common practice period
 * of music (no mode specified).
 */
public final class KeySignatures {
  /**
   * Disable constructor.
   */
  private KeySignatures() {
    throw new AssertionError("Instantiating static class");
  }

  public static final KeySignature SEVEN_FLAT = new KeySignature(
      List.of(FLAT, FLAT, FLAT, FLAT, FLAT, FLAT, FLAT), ScaleValue.C
  );
  public static final KeySignature SIX_FLAT = new KeySignature(
      List.of(FLAT, FLAT, FLAT, FLAT, FLAT, FLAT, NATURAL), ScaleValue.G
  );
  public static final KeySignature FIVE_FLAT = new KeySignature(
      List.of(FLAT, FLAT, NATURAL, FLAT, FLAT, FLAT, NATURAL), ScaleValue.D
  );
  public static final KeySignature FOUR_FLAT = new KeySignature(
      List.of(FLAT, FLAT, NATURAL, FLAT, FLAT, NATURAL, NATURAL), ScaleValue.A
  );
  public static final KeySignature THREE_FLAT = new KeySignature(
      List.of(FLAT, NATURAL, NATURAL, FLAT, FLAT, NATURAL, NATURAL), ScaleValue.E
  );
  public static final KeySignature TWO_FLAT = new KeySignature(
      List.of(FLAT, NATURAL, NATURAL, FLAT, NATURAL, NATURAL, NATURAL), ScaleValue.B
  );
  public static final KeySignature ONE_FLAT = new KeySignature(
      List.of(NATURAL, NATURAL, NATURAL, FLAT, NATURAL, NATURAL, NATURAL), ScaleValue.F
  );
  public static final KeySignature NO_ACCIDENTAL = new KeySignature(
      List.of(NATURAL, NATURAL, NATURAL, NATURAL, NATURAL, NATURAL, NATURAL), ScaleValue.C
  );
  public static final KeySignature ONE_SHARP = new KeySignature(
      List.of(NATURAL, NATURAL, NATURAL, NATURAL, NATURAL, NATURAL, SHARP), ScaleValue.G
  );
  public static final KeySignature TWO_SHARP = new KeySignature(
      List.of(NATURAL, NATURAL, SHARP, NATURAL, NATURAL, NATURAL, SHARP), ScaleValue.D
  );
  public static final KeySignature THREE_SHARP = new KeySignature(
      List.of(NATURAL, NATURAL, SHARP, NATURAL, NATURAL, SHARP, SHARP), ScaleValue.A
  );
  public static final KeySignature FOUR_SHARP = new KeySignature(
      List.of(NATURAL, SHARP, SHARP, NATURAL, NATURAL, SHARP, SHARP), ScaleValue.E
  );
  public static final KeySignature FIVE_SHARP = new KeySignature(
      List.of(NATURAL, SHARP, SHARP, NATURAL, SHARP, SHARP, SHARP), ScaleValue.B
  );

  public static final KeySignature SIX_SHARP = new KeySignature(
      List.of(SHARP, SHARP, SHARP, NATURAL, SHARP, SHARP, SHARP), ScaleValue.F
  );
  public static final KeySignature SEVEN_SHARP = new KeySignature(
      List.of(SHARP, SHARP, SHARP, SHARP, SHARP, SHARP, SHARP), ScaleValue.C
  );

  private static final List<KeySignature> ALL_KEY_SIGNATURES = new ArrayList<>(Arrays.asList(
      SEVEN_FLAT, SIX_FLAT, FIVE_FLAT, FOUR_FLAT, THREE_FLAT, TWO_FLAT, ONE_FLAT, NO_ACCIDENTAL,
      SEVEN_SHARP, SIX_SHARP, FIVE_SHARP, FOUR_SHARP, THREE_SHARP, TWO_SHARP, ONE_SHARP
  ));

  public static KeySignature getByNoteAtDegree(
      ScaleValue scaleValue, Accidental accidental, int degree
  ) {
    for (KeySignature key : ALL_KEY_SIGNATURES) {
      List<ScaleValue> noteNameScale = key.getScaleLetterNames(key.getIonianTonic());
      List<Accidental> keyAccidentals = key.getAccidentals();

      if (noteNameScale.get(degree) == scaleValue
          && keyAccidentals.get(degree) == accidental) {
        return key;
      }
    }
    throw new InvalidScaleException("Key signature not found");
  }
}
